ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc

```diff
@@ -0,0 +1,268 @@
+}
+
+/**
+ * @brief Configure llama2c instance
+ */
+void
+TensorFilterLlama2c::configure_instance (const GstTensorFilterProperties *prop)
+{
+  // build the Transformer via the model file
+  build_transformer (transformer, (char *) prop->model_files[0]);
+  config = &transformer->config;
+
+  if (steps == 0 || steps > config->seq_len)
+    steps = config->seq_len; // override to ~max length
+
+  // build the Tokenizer via the tokenizer file
+  build_tokenizer (tokenizer, (char *) prop->model_files[1], config->vocab_size);
+
+  // build the Sampler
+  build_sampler (sampler, config->vocab_size, TEMPERATURE, TOPP, RNG_SEED);
+}
+
+/**
+ * @brief Invoke llama2c using input tensors
+ */
+void
+TensorFilterLlama2c::invoke (const GstTensorMemory *input, GstTensorMemory *output)
+{
+  char *result = NULL, *prompt;
+  int len;
+
+  prompt = (char *) input[0].data;
+  len = strlen (prompt);
+
+  if (len > 1) {
+    prompt[strlen (prompt) - 1] = ' '; // Remove new line from text file
+  } else {
+    prompt = NULL;
+  }
+
+  output->size = tokenizer->max_token_length * steps;
+
+  result = get_tokens (transformer, tokenizer, sampler, prompt, steps);
+
+  output[0].data = result;
+  output[0].size = strlen (result);
+}
+
+/**
+ * @brief Get llama2c framework info.
+ */
+void
+TensorFilterLlama2c::getFrameworkInfo (GstTensorFilterFrameworkInfo &info)
+{
+  info.name = name;
+  info.allow_in_place = FALSE;
+  info.allocate_in_invoke = TRUE;
+  info.run_without_model = FALSE;
+  info.verify_model_path = TRUE;
+}
+
+/**
+ * @brief Get llama2c model info.
+ */
+int
+TensorFilterLlama2c::getModelInfo (
+    model_info_ops ops, GstTensorsInfo &in_info, GstTensorsInfo &out_info)
+{
+  switch (ops) {
+    case GET_IN_OUT_INFO:
+      in_info.num_tensors = 1;
+      in_info.format = _NNS_TENSOR_FORMAT_FLEXIBLE;
+
+      out_info.num_tensors = 1;
+      out_info.format = _NNS_TENSOR_FORMAT_STATIC;
+      out_info.info[0].type = _NNS_UINT8;
+      out_info.info[0].dimension[0] = tokenizer->max_token_length * steps;
+      break;
+    case SET_INPUT_INFO:
+      break;
+    default:
+      return -ENOENT;
+  }
+
+  return 0;
+}
+
+/**
+ * @brief Method to handle the event
+ */
+int
+TensorFilterLlama2c::eventHandler (event_ops ops, GstTensorFilterFrameworkEventData &data)
+{
+  UNUSED (ops);
+  UNUSED (data);
+  return 0;
+}
+
+/** @brief Initialize this object for tensor_filter subplugin runtime register */
+void
+TensorFilterLlama2c::init_filter_llama2c ()
+{
+  transformer = (Transformer *) malloc (sizeof (Transformer));
+  tokenizer = (Tokenizer *) malloc (sizeof (Tokenizer));
+  sampler = (Sampler *) malloc (sizeof (Sampler));
+  registered = tensor_filter_subplugin::register_subplugin<TensorFilterLlama2c> ();
+}
+
+/** @brief Destruct the subplugin */
+void
+TensorFilterLlama2c::fini_filter_llama2c ()
+{
+  free_transformer (transformer);
+  free_tokenizer (tokenizer);
+  free_sampler (sampler);
+
+  free (transformer);
+  free (tokenizer);
+  free (sampler);
```

Comment:
free(NULL) is safe to use without any side effects, but if a valid pointer is passed to free(), there seems to be a risk of a double free.
How about assigning nullptr after free?
```
free (transformer);
transformer  = nullptr;

free (tokenizer);
tokenizer = nullptr;

free (sampler);
sampler = nullptr;
```

Reply:
Assigned nullptr after free
Thank you :)

Reply:
Thank you.